### ISSUE TITLE
Excessive battery drain even when desktop is off

### ISSUE THREAD

**Issue by dwtokarski** [maruos/maruos#89]
Hi,

I have been trying out MaruOS v0.4.1 on a Nexus 7 (2013 2nd ed., flo).

Functionally everything works fine. However, even when the desktop is turned off there is considerable battery drain--like 25% down overnight--far in excess of any other firmware I've run on this hardware.

I've been using an app called System Monitor from NextApp to try and tease out what's causing this but not seeing anything definitive. It logs a steady decline in charge with a corresponding steady discharge rate (no spikes or unexplained variations) even during periods of no CPU use. Inspecting the per task power use doesn't finger any culprits either. During idle periods the kernel seems to be the top power user, but it's hardly used any CPU time.

It's not a matter of the battery self-discharging in any way. After fully charging and verifying a 100% charge on the battery display, I've tried powering the tablet completely off and letting it sit for 24+ hours.  Once turned back on, the battery is still at 100%.  Then the high discharge rate resumes.

Lately I've reflashed the factory stock firmware, charged to 100%, and let the tablet sleep overnight. In the morning it's still showing 100%.  This is consistent with the behavior on a couple of other N7s I have, one running CM11 and another running LineageOS 14.1.

My wild and uninformed speculation is that maybe it has something to do with MaruOS polling the connected state of the slimport too frequently, and maybe that's keeping the kernel from entering any deeper sleep state that it might use if it were truly idle.

Thoughts?  Any experiments I can try  to help pin this down?
 
Thanks!

**Comment by pdsouza** [comment]
Thank you for the detailed report! I'm working with our N7 maintainer @TMartinPPC to look into this.

By the way, since you mention polling slimport state--are you leaving the N7 connected to an external display over slimport when experiencing this drain or is it all the time even when disconnected? The way we detect an HDMI display is connected is by registering a display listener with the Android framework's display server and receiving a callback so it's unlikely that this is the cause.

**Comment by dwtokarski** [comment]
It's completely disconnected from anything--just sitting there on a table top all by itself, screen dark and apparently sleeping.

There could still be something going on with connection detection depending on what Android does under the hood when you register that listener.  One way for that to work would be if cable insertion caused a hardware interrupt to the CPU which then wakes it  from a low power sleep state.  Another would be for a periodic timer interrupt to wake the CPU many times per second so the listener or lower level code can check port status for the presence of a connection.  I suppose a third possibility could be that just having the listener active precludes deep sleep for some reason--could even be an Android bug.

Is there anything I can do from a command line like kill a process, or change a hidden option someplace, which would de-register your listener and make the system oblivious to cable/monitor connections?  It would be interesting to do that and see if anything changes.

You support a phone platform too, right?  How's the battery discharge rate on that? You might need to put it in airplane mode to approximate an N7 with wifi off.

**Comment by TMartinPPC** [comment]
I tried to reproduce this last night but was not successful. I charged my Nexus 7 to 100% and then unplugged it prior to going to bed. This is what I woke up to. I might try to do this again tonight to see if my results vary.

Not sure if it makes a difference but here's some info on my install. I'm on the latest Maru version 0.4.1. I'm using an unofficial TWRP version 3.1.1-0 from followmsi on xda. I installed open_gapps_arm-6.0-pico-20170601.zip and have installed several other Google apps from the Play store afterward.

![screenshot_20170928-073401](https://user-images.githubusercontent.com/1407126/30975808-30489db6-a441-11e7-84d9-811d813417f0.png)

![screenshot_20170928-073427](https://user-images.githubusercontent.com/1407126/30975816-3357395e-a441-11e7-8699-0053893e4274.png)

**Comment by dwtokarski** [comment]
Hi,

Sorry for the long delay, but the test cycle time for this problem is hours for each trial.  I've confirmed the problem goes away with an absolutely clean install, and comes back if I use TWRP to restore my full configuration from backup.  So at least I've had a known though complex configuration to experiment with.

@TMartinPPC, I have TWRP 3.1.1-0 installed which I then used to install maru-v0.4.1-update-flo-59edff30.zip, open_gapps-arm-6.0-nano-20170914.zip, and SuperSU-v2.82-201705271822.zip.  Like you, I have a number of additional Play Store apps installed, but no where near as many as I have on other N7s, and none which cause problems anywhere else.

Or so I thought.  Eventually it hit me that there _is_ one app on this tablet that I've not tried elsewhere--the substratum theming engine with the legacy inversionUI theme.  Looks nice, but it was the problem somehow.  After carefully and completely removing substratum from the tablet I see a power use profile just as good as any other N7 / ROM configuration.

So cancel the mayday.

I do wish I understood the mechanism for how the theme engine kills battery life, but at least this is a solution.

And MaruOS rocks. I'm looking forward to using it.

Cheers!

--Dennis
